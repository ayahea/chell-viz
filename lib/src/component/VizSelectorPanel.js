"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var semantic_ui_react_1 = require("semantic-ui-react");
var chell_1 = require("chell");
var NGLComponent_1 = require("../component/NGLComponent");
var SpringComponent_1 = require("../component/SpringComponent");
var TComponent_1 = require("../component/TComponent");
var ReactHelper_1 = require("../helper/ReactHelper");
var ContactMap_1 = require("./ContactMap");
var defaultProps = {
    data: {},
    height: 450,
    initialViz: chell_1.VIZ_TYPE['T-SNE'],
    onDataSelect: function (e) {
        return;
    },
    padding: 15,
    selectedData: undefined,
    supportedVisualizations: [],
    width: 450,
};
var initialState = {
    selectedViz: defaultProps.initialViz,
};
exports.VizSelectorPanel = ReactHelper_1.withDefaultProps(defaultProps, /** @class */ (function (_super) {
    __extends(VizSelectorPanelClass, _super);
    function VizSelectorPanelClass(props) {
        var _this = _super.call(this, props) || this;
        _this.state = initialState;
        _this.onVizSelect = function (event, data) {
            _this.setState({
                selectedViz: data.value,
            });
        };
        _this.generateDropdownItems = function (supportedVisualizations) {
            return supportedVisualizations
                .map(function (viz) { return ({
                key: viz,
                text: viz,
                value: viz,
            }); })
                .sort(function (a, b) { return a.key.localeCompare(b.key); });
        };
        _this.state = {
            selectedViz: props.initialViz ? props.initialViz : chell_1.VIZ_TYPE.SPRING,
        };
        return _this;
    }
    VizSelectorPanelClass.prototype.render = function () {
        // N.B. We are only setting the width of the VizSelectorPanel, explicitly leaving out the height.
        // This means a component can only grow vertically, but not horizontally, and be correctly styled in containers.
        var _a = this.props, data = _a.data, selectedData = _a.selectedData, supportedVisualizations = _a.supportedVisualizations, width = _a.width;
        return (React.createElement("div", { className: "VizSelectorPanel", style: { width: width } },
            React.createElement(semantic_ui_react_1.Dropdown, { options: this.generateDropdownItems(supportedVisualizations), fluid: true, onChange: this.onVizSelect, defaultValue: this.props.initialViz }),
            React.createElement(semantic_ui_react_1.Card, { fluid: true, raised: true }, this.renderVizContainer(this.state.selectedViz, data, selectedData))));
    };
    VizSelectorPanelClass.prototype.renderVizContainer = function (viz, data, selectedData) {
        var padding = this.props.padding;
        var paddedHeight = this.props.height - padding * 2;
        var paddedWidth = this.props.width - padding * 2;
        switch (viz) {
            case chell_1.VIZ_TYPE['T-SNE']:
                return (React.createElement(TComponent_1.TComponent, { data: data['T-SNE'], height: paddedHeight, padding: padding, width: paddedWidth }));
            case chell_1.VIZ_TYPE.SPRING:
                return (React.createElement(SpringComponent_1.SpringComponent, { data: data.Spring, height: paddedHeight, padding: padding, width: paddedWidth }));
            case chell_1.VIZ_TYPE.NGL:
                return (React.createElement(NGLComponent_1.NGLComponent, { data: data.NGL, height: paddedHeight, padding: padding, width: paddedWidth }));
            case chell_1.VIZ_TYPE.CONTACT_MAP:
                return (React.createElement(ContactMap_1.ContactMap, { data: data['Contact Map'], enableSliders: true, height: paddedHeight, onMouseEnter: this.props.onDataSelect, padding: padding, selectedData: selectedData, width: paddedWidth }));
            default:
                throw new Error("Unknown viz: " + viz);
        }
    };
    return VizSelectorPanelClass;
}(React.Component)));
