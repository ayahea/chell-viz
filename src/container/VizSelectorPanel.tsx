import * as React from 'react';

import { Dropdown, DropdownItemProps } from 'semantic-ui-react';

import { ProteinViewer } from './ProteinViewer';
import { SpringContainer } from './SpringContainer';
import { TContainer } from './TContainer';

export enum VIZ_TYPE {
  NGL = 'NGL',
  SPRING = 'SPRING',
  'T-SNE' = 'T-SNE',
}

export interface IVizSelectorPanelProps {
  initialViz?: VIZ_TYPE;
}

export interface IVizSelectorPanelState {
  selectedViz: VIZ_TYPE;
}

export class VizSelectorPanel extends React.Component<IVizSelectorPanelProps, IVizSelectorPanelState> {
  public static defaultParams: Partial<IVizSelectorPanelProps> = {
    initialViz: VIZ_TYPE['T-SNE'],
  };

  protected dropdownItems: DropdownItemProps[] = Object.keys(VIZ_TYPE).map(viz => ({
    key: viz,
    text: viz,
    value: viz,
  }));

  constructor(props: IVizSelectorPanelProps) {
    super(props);
    this.state = {
      selectedViz: props.initialViz ? props.initialViz : VIZ_TYPE.SPRING,
    };
  }

  public render() {
    return (
      <div className="VisSelectorPanel">
        <Dropdown
          options={this.dropdownItems}
          placeholder={'Select a Visualization!'}
          fluid={true}
          onChange={this.onVizSelect}
        />
        {this.renderVizContainer(this.state.selectedViz)}
      </div>
    );
  }

  protected renderVizContainer(viz: VIZ_TYPE) {
    const dataDir = 'centroids';
    switch (viz) {
      case VIZ_TYPE['T-SNE']:
        return <TContainer dataDir={dataDir} />;
      case VIZ_TYPE.SPRING:
        return <SpringContainer dataDir={dataDir} />;
      case VIZ_TYPE.NGL:
        return <ProteinViewer />;
      default:
        throw new Error(`Unknown viz: ${viz}`);
    }
  }
  protected onVizSelect = (event: React.SyntheticEvent<any>, data: any) => {
    this.setState({
      selectedViz: data.value,
    });
  };
}
