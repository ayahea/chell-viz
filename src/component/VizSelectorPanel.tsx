import * as React from 'react';
import { Card, Dropdown, DropdownItemProps } from 'semantic-ui-react';
import { ISpringGraphData } from 'spring';
import {
  CHELL_DATA_TYPE,
  IContactMapData,
  ICouplingScore,
  NGL_DATA_TYPE,
  T_SNE_DATA_TYPE,
  VIZ_TYPE,
} from '../../types/chell';
import { NGLComponent } from '../component/NGLComponent';
import { SpringComponent } from '../component/SpringComponent';
import { TComponent } from '../component/TComponent';
import { withDefaultProps } from '../helper/ReactHelper';
import { ContactMap } from './ContactMap';

const defaultProps = {
  data: {} as Partial<{ [K in VIZ_TYPE]: CHELL_DATA_TYPE }>,
  height: 450,
  initialViz: VIZ_TYPE['T-SNE'],
  onDataSelect: (e: any) => {
    return;
  },
  padding: 15,
  selectedData: undefined as ICouplingScore | undefined,
  supportedVisualizations: [] as VIZ_TYPE[],
  width: 450,
};

const initialState = {
  selectedViz: defaultProps.initialViz,
};

type Props = {} & typeof defaultProps;
type State = Readonly<typeof initialState>;

export const VizSelectorPanel = withDefaultProps(
  defaultProps,

  /**
   * A single visualization panel allowing a user to select how they wish to view data.
   *
   * @export
   * @extends {React.Component<Props, State>}
   */
  class VizSelectorPanelClass extends React.Component<Props, State> {
    public readonly state: State = initialState;

    constructor(props: Props) {
      super(props);
      this.state = {
        selectedViz: props.initialViz ? props.initialViz : VIZ_TYPE.SPRING,
      };
    }

    public render() {
      // N.B. We are only setting the width of the VizSelectorPanel, explicitly leaving out the height.
      // This means a component can only grow vertically, but not horizontally, and be correctly styled in containers.
      const { data, selectedData, supportedVisualizations, width } = this.props;

      return (
        <div className="VizSelectorPanel" style={{ width }}>
          <Dropdown
            options={this.generateDropdownItems(supportedVisualizations)}
            fluid={true}
            onChange={this.onVizSelect}
            defaultValue={this.props.initialViz}
          />
          {
            <Card fluid={true} raised={true}>
              {this.renderVizContainer(this.state.selectedViz, data, selectedData)}
            </Card>
          }
        </div>
      );
    }

    protected renderVizContainer(
      viz: VIZ_TYPE,
      data: Partial<{ [K in VIZ_TYPE]: CHELL_DATA_TYPE }>,
      selectedData?: ICouplingScore | number,
    ) {
      const { padding } = this.props;
      const paddedHeight = this.props.height - padding * 2;
      const paddedWidth = this.props.width - padding * 2;
      switch (viz) {
        case VIZ_TYPE['T-SNE']:
          return (
            <TComponent
              data={data['T-SNE'] as T_SNE_DATA_TYPE}
              height={paddedHeight}
              padding={padding}
              width={paddedWidth}
            />
          );
        case VIZ_TYPE.SPRING:
          return (
            <SpringComponent
              data={data.Spring as ISpringGraphData}
              height={paddedHeight}
              padding={padding}
              width={paddedWidth}
            />
          );
        case VIZ_TYPE.NGL:
          return (
            <NGLComponent
              data={data.NGL as NGL_DATA_TYPE}
              height={paddedHeight}
              padding={padding}
              width={paddedWidth}
            />
          );
        case VIZ_TYPE.CONTACT_MAP:
          return (
            <ContactMap
              data={data['Contact Map'] as IContactMapData}
              enableSliders={true}
              height={paddedHeight}
              onMouseEnter={this.props.onDataSelect}
              padding={padding}
              selectedData={selectedData as number}
              width={paddedWidth}
            />
          );
        default:
          throw new Error(`Unknown viz: ${viz}`);
      }
    }

    protected onVizSelect = (event: React.SyntheticEvent<any>, data: any) => {
      this.setState({
        selectedViz: data.value,
      });
    };

    protected generateDropdownItems = (supportedVisualizations: VIZ_TYPE[]): DropdownItemProps[] => {
      return supportedVisualizations
        .map(viz => ({
          key: viz,
          text: viz,
          value: viz,
        }))
        .sort((a, b) => a.key.localeCompare(b.key));
    };
  },
);
