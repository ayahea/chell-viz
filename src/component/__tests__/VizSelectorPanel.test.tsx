import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';

import { VIZ_TYPE, VizSelectorPanel } from '../VizSelectorPanel';

describe('VizSelectorPanel', () => {
  const testDataDir = 'sentence/kongsvinger';
  const testData = {};

  test('Should match existing snapshot when given no initial visualization.', () => {
    expect(toJson(shallow(<VizSelectorPanel data={testData} />))).toMatchSnapshot();
  });

  Object.keys(VIZ_TYPE).forEach(vizKey => {
    const viz = VIZ_TYPE[vizKey];
    test(`Should match existing snapshot when given initial viz of ${viz}.`, () => {
      expect(toJson(shallow(<VizSelectorPanel data={testData} initialViz={viz} />))).toMatchSnapshot();
    });
  });

  test('Should throw an error when given an invalid visualization option.', () => {
    // Testing a render error is slightly different - here we are checking "if I were to render this, would it throw an error?".
    expect(() =>
      shallow(<VizSelectorPanel initialViz={'Through the Looking Glass' as VIZ_TYPE} data={testData} />),
    ).toThrow();
  });
});
