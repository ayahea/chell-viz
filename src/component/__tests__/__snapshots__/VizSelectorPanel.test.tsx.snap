// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VizSelectorPanel Should match existing snapshot when given initial viz of Contact Map. 1`] = `
<div
  className="VizSelectorPanel"
  style={
    Object {
      "width": 450,
    }
  }
>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    closeOnBlur={true}
    deburr={false}
    defaultValue="Contact Map"
    fluid={true}
    icon="dropdown"
    minCharacters={1}
    noResultsMessage="No results found."
    onChange={[Function]}
    openOnFocus={true}
    options={Array []}
    renderLabel={[Function]}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    wrapSelection={true}
  />
  <Card
    fluid={true}
    raised={true}
  >
    <ContactMapComponentClass
      contactColor="#009999"
      couplingColor="#000000"
      data={
        Object {
          "contactMonomer": Array [],
          "couplingScore": Array [],
          "distanceMapMonomer": Array [],
        }
      }
      height={405}
      highlightColor="#0000ff"
      onMouseEnter={[Function]}
      width={405}
    />
  </Card>
</div>
`;

exports[`VizSelectorPanel Should match existing snapshot when given initial viz of NGL. 1`] = `
<div
  className="VizSelectorPanel"
  style={
    Object {
      "width": 450,
    }
  }
>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    closeOnBlur={true}
    deburr={false}
    defaultValue="NGL"
    fluid={true}
    icon="dropdown"
    minCharacters={1}
    noResultsMessage="No results found."
    onChange={[Function]}
    openOnFocus={true}
    options={Array []}
    renderLabel={[Function]}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    wrapSelection={true}
  />
  <Card
    fluid={true}
    raised={true}
  >
    <Component />
  </Card>
</div>
`;

exports[`VizSelectorPanel Should match existing snapshot when given initial viz of Spring. 1`] = `
<div
  className="VizSelectorPanel"
  style={
    Object {
      "width": 450,
    }
  }
>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    closeOnBlur={true}
    deburr={false}
    defaultValue="Spring"
    fluid={true}
    icon="dropdown"
    minCharacters={1}
    noResultsMessage="No results found."
    onChange={[Function]}
    openOnFocus={true}
    options={Array []}
    renderLabel={[Function]}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    wrapSelection={true}
  />
  <Card
    fluid={true}
    raised={true}
  >
    <SpringComponentClass
      canvasBackgroundColor={13421772}
      data={
        Object {
          "links": Array [],
          "nodes": Array [],
        }
      }
      height={450}
      selectedCategory=""
      width={450}
    />
  </Card>
</div>
`;

exports[`VizSelectorPanel Should match existing snapshot when given initial viz of T-SNE. 1`] = `
<div
  className="VizSelectorPanel"
  style={
    Object {
      "width": 450,
    }
  }
>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    closeOnBlur={true}
    deburr={false}
    defaultValue="T-SNE"
    fluid={true}
    icon="dropdown"
    minCharacters={1}
    noResultsMessage="No results found."
    onChange={[Function]}
    openOnFocus={true}
    options={Array []}
    renderLabel={[Function]}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    wrapSelection={true}
  />
  <Card
    fluid={true}
    raised={true}
  >
    <TComponentClass
      data={
        Array [
          Array [
            0,
          ],
          Array [
            0,
          ],
        ]
      }
      height={405}
      pointColor="#000000"
      width={405}
    />
  </Card>
</div>
`;

exports[`VizSelectorPanel Should match existing snapshot when given no initial visualization. 1`] = `
<div
  className="VizSelectorPanel"
  style={
    Object {
      "width": 450,
    }
  }
>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    closeOnBlur={true}
    deburr={false}
    defaultValue="T-SNE"
    fluid={true}
    icon="dropdown"
    minCharacters={1}
    noResultsMessage="No results found."
    onChange={[Function]}
    openOnFocus={true}
    options={Array []}
    renderLabel={[Function]}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    wrapSelection={true}
  />
  <Card
    fluid={true}
    raised={true}
  >
    <TComponentClass
      data={
        Array [
          Array [
            0,
          ],
          Array [
            0,
          ],
        ]
      }
      height={405}
      pointColor="#000000"
      width={405}
    />
  </Card>
</div>
`;
